mindmeld/text_preparation/tokenizers.py

```diff
@@ -0,0 +1,258 @@
+
+"""This module contains Tokenizers."""
+
+from abc import ABC, abstractmethod
+import logging
+import unicodedata
+
+from .spacy_model_factory import SpacyModelFactory
+from ..constants import UNICODE_NON_LATIN_CATEGORY, UNICODE_SPACE_CATEGORY
+
+logger = logging.getLogger(__name__)
+
+
+class Tokenizer(ABC):
+    """Abstract Tokenizer Base Class."""
+
+    def __init__(self):
+        """Creates a tokenizer instance."""
+        pass
+
+    @abstractmethod
+    def tokenize(self, text):
+        """
+        Args:
+            text (str): Input text.
+        Returns:
+            tokens (List[str]): List of tokens.
+        """
+        raise NotImplementedError("Subclasses must implement this method")
+
+
+class NoOpTokenizer(Tokenizer):
+    """A No-Ops tokenizer."""
+
+    def __init__(self):
+        """Initialize the NoOpTokenizer."""
+        pass
+
+    def tokenize(self, text):
+        """
+        Args:
+            text (str): Input text.
+        Returns:
+            tokens (List[str]): List of tokens.
+        """
+        return [text]
+
+
+class LetterTokenizer(Tokenizer):
+    """A Tokenizer that splits text at the character level."""
+
+    def __init__(self):
+        """Initializes the LetterTokenizer."""
+        pass
+
+    def tokenize(self, text):
+        """
+        Identify tokens in text and create normalized tokens that contain the text and start index.
+        Args:
+            text (str): The text to normalize
+        Returns:
+            tokens (List[Dict]): List of tokenized tokens which a represented as dictionaries.
+                Keys include "start" (token starting index), "end" (token ending index), and
+                "text" (token text). For example: [{"start": 0, "text":"hello", "end":4}]
+        """
+        actions_by_char = LetterTokenizer.get_actions_by_char(text)
+        token_num_by_char = LetterTokenizer.get_token_num_by_char(actions_by_char)
+        return LetterTokenizer.create_tokens(text, token_num_by_char)
+
+    @staticmethod
+    def get_actions_by_char(text):
+        """
+        Identify the appropriate action at each character ("separate", "skip", "no_action")
+        Actions are determined by the unicode category of each character. More details can be
+        found here: https://www.compart.com/en/unicode/category.
+        Args:
+            text (str): The text to process and get actions per character.
+        Returns:
+            actions_by_char (List[str]): List of actions corresponding to each character.
+                For example: ["separate", "separate", "skip", "no_action"]
+        """
+        category_by_char = [unicodedata.category(x) for x in text]
+        actions_by_char = []
+        for index, category in enumerate(category_by_char):
+            same_category_as_previous = (
+                index > 0 and category[0] == category_by_char[index - 1][0]
+            )
+            if category == UNICODE_SPACE_CATEGORY:
+                action = "skip"
```

Comment:
I think we can add a enum class for these actions in the `LetterTokenizer `

Reply:
After combining the `get_actions_by_char` and `get_token_num_by_char` functions together as suggested, we no longer need to maintain any representation of "actions".